### DISCUSSION TITLE
How to make a REST API call to the minder server

### DISCUSSION THREAD

**Discussion by AGMETEOR** [stacklok/minder#1957] [General]
From the minder.proto file, I see it's possible to make REST based API call instead instead of GRPC one. However, I am failing to authenticate to one of the endpoints at `/api/v1/auth/url`. Here's my cuRL command
```
curl --location --request GET 'http://api.stacklok.com/api/v1/auth/url' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <access_token>' \
--data '{
    "provider": "github",
    "cli": true,
    "port": 3000,
    "project_id": "",
    "owner": ""
}'
```

This still says "unauthenticated"

**Comment by JAORMX** [comment]
Hey! I see how that endpoint can be confusing. Minder doesn't do authentication by itself. Instead, we're leveraging keycloak to do the heavy lifting, and we subsequently just verify the token that keycloak provides. If you're interested in how that works, it's an OAuth2 flow that can be seen here https://github.com/stacklok/minder/blob/main/cmd/cli/app/auth/auth_login.go

**Comment by JAORMX** [comment]
Also, regarding that call, you'll need to provide a project ID. Which you get once you register.

**Comment by AGMETEOR** [comment]
@JAORMX put the project id but still says unauthenticated.
```
curl --location --request GET 'http://api.stacklok.com/api/v1/auth/url' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <access_token>' \
--data '{
    "provider": "github",
    "cli": true,
    "port": 3000,
    "project_id": <my_proj_id>,
    "owner": ""
}'
```

**Reply by JAORMX** [reply]
Is it on that specific endpoint? Or are other endpoints behaving the same?

**Reply by AGMETEOR** [reply]
I have not checked others, was still trying to get this one to work.

**Reply by evankanderson** [reply]
[Looking at `AuthUnaryInterceptor`](https://github.com/stacklok/minder/blob/main/internal/controlplane/handlers_authz.go#L152), I notice [it is calling `gauth.AuthFromMD(...)`](https://github.com/stacklok/minder/blob/main/internal/controlplane/handlers_authz.go#L170) to get the request token.  Looking at [the documentation for `AuthFromMD`](https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware/v2@v2.0.1/interceptors/auth#AuthFromMD), i wonder if that method doesn't look in the HTTP `Authorization` header for HTTP gateway requests.

**Reply by AGMETEOR** [reply]
Looks like it's the case. I also tried another endpoint and still unauthenticated
```
curl --location --request GET 'http://api.stacklok.com/api/v1/repositories/provider/github' \
--header 'authorization: bearer <access_token>' \
--header 'Content-Type: application/json' \
--data '{
    "provider": "github",
    "cli": true,
    "port": 3000,
    "project_id": "<project_id>",
    "owner": ""
}'
```

**Reply by evankanderson** [reply]
https://stackoverflow.com/a/49072934/1543502 seems related -- it seems like the GRPC gateway is not passing the authorization header.

**Reply by evankanderson** [reply]
It looks like we can use [`WithIncomingHeaderMatcher`](https://pkg.go.dev/github.com/grpc-ecosystem/grpc-gateway/runtime#WithIncomingHeaderMatcher) on our `ServeMux` if we want to customize this.

https://github.com/grpc-ecosystem/grpc-gateway/blob/v1.16.0/runtime/mux.go#L73 seems to be the default header matcher.

**Reply by evankanderson** [reply]
It looks like `Authorization` is going to be mapped to `grpcgateway-Authorization` (because why let a little thing like canonicalization stand in your way  😁 ).

We can either adjust the `ServeMux` to map this to the GRPC `authorization` metadata, or we can adjust the auth handler to look in both places for the data.  I'm inclined to adjust the `ServeMux`, but I'd love to hear other opinions.

**Reply by evankanderson** [reply]
... but then I find this: https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/context.go#L152

If you're willing to do some testing locally, you might try logging the token you receive on a local build, but you definitely _don't_ want to push that off your local machine and start logging auth tokens elsewhere.

**Reply by evankanderson** [reply]
https://github.com/stacklok/minder/pull/1960 should provide us some additional log message details on the client.

**Reply by AGMETEOR** [reply]
Ok I'll try setting up a local server and doing some testing with the token there. I'll let you know how that goes.

**Reply by AGMETEOR** [reply]
So I am trying to run the server locally but the container `minder-minder-1` keeps failing and restarting. The logs I see an error
```
Error executing root command: unable to create server: failed to create crypto engine: failed to read token key file: failed to read key: open /app/.ssh/token_key_passphrase: no such file or directory
```

**Reply by AGMETEOR** [reply]
Never mind fixed with `openssl rand -base64 32 > .ssh/token_key_passphrase`

**Reply by AGMETEOR** [reply]
Ok so I have an update. The interceptor picks up the token as expected. The error was occurring because of a failure to parse the token. And the reason for the malformed token was my fault, I opened the credentials.json file with my editor and while copying I introduced tab/space? characters in the token.

Otherwise, I have tested on both `api.stacklok.com` and `localhost` and authentication over REST happens perfectly fine.

Thanks a lot @evankanderson and @JAORMX

**Reply by evankanderson** [reply]
Excellent!  Glad to hear it was a misplaced character in the file.

Hopefully #1960 should make future such errors easier to catch.

We should also think about having different credential files for each grpc endpoint (to make it easier to switch between servers).

**Reply by AGMETEOR** [reply]
Yeah separating credential files is a good idea. Created the issue here https://github.com/stacklok/minder/issues/1961
I can explore ways to do this.
